subscription.go

```diff
@@ -505,31 +533,35 @@ func (sc *SubscriptionClient) Unsubscribe(id string) error {
 	return sc.protocol.Unsubscribe(sc.context, id)
 }
 
-// Run start websocket client and subscriptions. If this function is run with goroutine, it can be stopped after closed
+// Run start the WebSocket client and subscriptions.
+// If the client is running, recalling this function will restart all registered subscriptions
+// If this function is run with goroutine, it can be stopped after closed
 func (sc *SubscriptionClient) Run() error {
+
+	sc.reset()
 	if err := sc.init(); err != nil {
 		return fmt.Errorf("retry timeout. exiting...")
 	}
 
 	sc.setIsRunning(true)
```

Comment:
I have noticed that subscriptions registered while `init()` are never sent to the server, due to silently skipping it until the subscription is set to running. Would you consider sending the unstarted subscriptions to the server, after `sc.setIsRunning(true)`?

Reply:
That is handled in the protocol. After authorizing with the server, the client has to wait until the server acknowledges it. Subscriptions are sent to the server before the acknowledgment is discarded.
https://github.com/hasura/go-graphql-client/blob/10471cf9d8df859e36584a8821e5ffc3d91c76b8/subscriptions_transport_ws.go#L178

Reply:
Thanks for this pointer! 

Then I see another synchronisation issue, due to using 2 different flags (`acknowledged`, `isRunning`) and using the `isRunning` for preventing subscription to be sent to the server, while the protocol sends subscriptions when `acknowledged` flag switches to true.

If only the `acknowledged` flag was used, then if a subscription is sent between the time a connection is acknowledged and the time the `init()` returns (currently marked as `isRunning=true`), it would successfully send the subscription to the server. Currently, there is a chance that the subscription is never sent, due to depending on another flag (`isRunning`) for submitting a subscription that was received before the `isRunning` flag is switched to true.

Reply:
I can see that `isRunning` was replaced by `clientStatus`. Just wondering if possible to couple clientStatus=running with acknowledged. This way, no synchronisation issue like the one described would be possible.

Reply:
updated. We only need to check the `acknowledge` status because the flag is only updated when the client is running

Reply:
g8, LGTM!